mars/services/scheduling/supervisor/speculation.py

```diff
@@ -0,0 +1,267 @@
+            "min_task_runtime", DEFAULT_SUBTASK_SPECULATION_MIN_TASK_RUNTIME
+        )
+        self._subtask_speculation_multiplier = speculation_config.get(
+            "multiplier", DEFAULT_SUBTASK_SPECULATION_MULTIPLIER
+        )
+        self._subtask_speculation_max_concurrent_run = speculation_config.get(
+            "max_concurrent_run", DEFAULT_SUBTASK_MAX_CONCURRENT_RUN
+        )
+        if self._subtask_speculation_enabled:
+            assert 1 >= self._subtask_speculation_threshold > 0
+            assert self._subtask_speculation_interval > 0
+            assert self._subtask_speculation_min_task_runtime > 0
+            assert self._subtask_speculation_multiplier > 0
+            assert self._subtask_speculation_max_concurrent_run > 0
+        self._speculation_execution_task = None
+
+    async def start(self):
+        if self._subtask_speculation_enabled:
+            self._speculation_execution_task = create_task_with_error_log(
+                self._speculative_execution()
+            )
+            logger.info(
+                "Speculative execution started with config %s.",
+                self._speculation_config,
+            )
+
+    async def stop(self):
+        if self._subtask_speculation_enabled:
+            self._speculation_execution_task.cancel()
+            await self._speculation_execution_task
+            logger.info("Speculative execution stopped.")
+
+    def add_subtask(self, subtask_info: SubtaskScheduleInfo):
+        # duplicate subtask add will be handled in `_speculative_execution`.
+        subtask = subtask_info.subtask
+        self._grouped_unfinished_subtasks[subtask.logic_key][
+            subtask.subtask_id
+        ] = subtask_info
+
+    def finish_subtask(self, subtask_info: SubtaskScheduleInfo):
+        subtask = subtask_info.subtask
+        grouped_finished_subtasks = self._grouped_finished_subtasks[subtask.logic_key]
+        grouped_finished_subtasks[subtask.subtask_id] = subtask_info
+        if len(grouped_finished_subtasks) == subtask.logic_parallelism:
+            self._grouped_finished_subtasks.pop(subtask.logic_key)
+            self._grouped_unfinished_subtasks.pop(subtask.logic_key, None)
+            logger.info(
+                "Subtask group with logic id %s parallelism %s finished.",
+                subtask.logic_key,
+                subtask.logic_parallelism,
+            )
+
+    async def _speculative_execution(self):
+        while True:
+            # check subtasks in the same group which has same logic key periodically, if some subtasks hasn't been
+            # finished in a considerably longer duration, then those subtasks maybe slow/hang subtasks, try resubmit
+            # it to other bands too.
+            await asyncio.sleep(self._subtask_speculation_interval)
+            for logic_id, subtask_infos_dict in dict(
+                self._grouped_finished_subtasks
+            ).items():
+                if subtask_infos_dict:
```

Comment:
To reduce indentation,

```suggestion
                if not subtask_infos_dict:
                    continue
```

Reply:
Good suggestion, I'll fix it